File: apps/web/lib/deal-enrichment.ts
import Anthropic from "@anthropic-ai/sdk";
import { getDbDeal, persistSourceChunks, upsertSource } from "./deal-db";
import { getSupabaseServiceClient } from "./supabase";
import { MODEL } from "./llm";
import type { DealFact } from "./types";

type EnrichedSource = {
  title: string;
  url: string;
  publisher?: string;
  snippet?: string;
};

type EnrichedFact = {
  factType: string;
  factValue: string;
  normalizedValue?: string;
  sourceUrl: string;
  evidenceText: string;
  confidence: number;
  reviewStatus?: DealFact["reviewStatus"];
};

type EnrichedConflict = {
  factType: string;
  values: string[];
  note: string;
};

type EnrichmentPayload = {
  sources?: EnrichedSource[];
  facts?: EnrichedFact[];
  conflicts?: EnrichedConflict[];
  remainingMissingFacts?: string[];
};

const SYSTEM = `You enrich private-equity deal facts using web search.
Use authoritative sources first: buyer, seller, target, investor, bank, law firm, regulator, reputable trade press.
Never invent undisclosed economics. Return strict JSON only. Low-confidence or uncertain facts must be review-gated.`;

export async function runDealEnrichment(externalDealId: string) {
  if (!process.env.ANTHROPIC_API_KEY) {
    throw new Error("Set ANTHROPIC_API_KEY to run web-search deal enrichment.");
  }

  const deal = await getDbDeal(externalDealId);
  if (!deal) throw new Error(`Unknown persisted deal: ${externalDealId}`);

  const supabase = getSupabaseServiceClient();
  const { data: dealRow, error: dealRowError } = await supabase
    .from("deals")
    .select("id")
    .eq("external_id", externalDealId)
    .single();
  if (dealRowError) throw new Error(dealRowError.message);
  const dealUuid = dealRow.id as string;

  const { data: run, error: runError } = await supabase
    .from("deal_enrichment_runs")
    .insert({
      trigger: "manual",
      status: "running",
      queries: deal.followUpSearches,
      metadata: { dealExternalId: externalDealId },
    })
    .select("id")
    .single();
  if (runError) throw new Error(runError.message);

  try {
    const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
    const response = await client.messages.create({
      model: MODEL,
      max_tokens: 3000,
      system: SYSTEM,
      tools: [{ type: "web_search_20250305", name: "web_search", max_uses: 8 }],
      messages: [
        {
          role: "user",
          content: `Deal: ${deal.name}
Theme: ${deal.theme}
Known facts:
${deal.facts.map((fact) => `- ${fact.factType}: ${fact.factValue}`).join("\n")}

Missing facts:
${deal.missingFacts.map((fact) => `- ${fact}`).join("\n")}

Run targeted searches using these query ideas:
${deal.followUpSearches.map((query) => `- ${query}`).join("\n")}

Return strict JSON:
{
  "sources": [{"title": string, "url": string, "publisher": string, "snippet": string}],
  "facts": [{"factType": string, "factValue": string, "normalizedValue": string, "sourceUrl": string, "evidenceText": string, "confidence": number, "reviewStatus": "verified" | "needs_review" | "not_disclosed"}],
  "conflicts": [{"factType": string, "values": string[], "note": string}],
  "remainingMissingFacts": string[]
}`,
        },
      ],
    });

    const text = response.content.map((part) => (part.type === "text" ? part.text : "")).join("");
    const payload = parseJson<EnrichmentPayload>(text) ?? {};
    const sourceIdByUrl = new Map<string, string>();

    for (const source of payload.sources ?? []) {
      const sourceId = await upsertSource({
        title: source.title,
        url: source.url,
        publisher: source.publisher,
        sourceType: "web_enrichment",
        rawText: source.snippet,
        metadata: { dealExternalId: externalDealId },
      });
      sourceIdByUrl.set(source.url, sourceId);
      if (source.snippet) {
        await persistSourceChunks(sourceId, source.snippet, {
          dealExternalId: externalDealId,
          enrichmentRunId: run.id,
        });
      }
    }

    const factRows = [];
    for (const fact of payload.facts ?? []) {
      const sourceId =
        sourceIdByUrl.get(fact.sourceUrl) ??
        (fact.sourceUrl
          ? await upsertSource({
              title: fact.sourceUrl,
              url: fact.sourceUrl,
              sourceType: "web_enrichment",
              rawText: fact.evidenceText,
              metadata: { dealExternalId: externalDealId },
            })
          : undefined);
      factRows.push({
        deal_id: dealUuid,
        fact_type: fact.factType,
        fact_value: fact.factValue,
        normalized_value: fact.normalizedValue ?? null,
        source_id: sourceId ?? null,
        evidence_text: fact.evidenceText,
        confidence: Math.max(0, Math.min(1, fact.confidence ?? 0.7)),
        extraction_method: "web_search",
        review_status: fact.reviewStatus ?? "needs_review",
      });
    }

    if (factRows.length) {
      const { error } = await supabase.from("deal_facts").insert(factRows);
      if (error) throw new Error(error.message);
    }

    const conflictRows = (payload.conflicts ?? []).map((conflict) => ({
      deal_id: dealUuid,
      fact_type: conflict.factType,
      values: conflict.values,
      note: conflict.note,
    }));
    if (conflictRows.length) {
      const { error } = await supabase.from("deal_fact_conflicts").insert(conflictRows);
      if (error) throw new Error(error.message);
    }

    const { error: updateDealError } = await supabase
      .from("deals")
      .update({
        missing_facts: payload.remainingMissingFacts ?? deal.missingFacts,
        updated_at: new Date().toISOString(),
      })
      .eq("external_id", externalDealId);
    if (updateDealError) throw new Error(updateDealError.message);

    const { error: completeError } = await supabase
      .from("deal_enrichment_runs")
      .update({
        status: "completed",
        sources_found: payload.sources?.length ?? 0,
        facts_created: factRows.length,
        completed_at: new Date().toISOString(),
      })
      .eq("id", run.id);
    if (completeError) throw new Error(completeError.message);

    return {
      runId: run.id as string,
      sourcesFound: payload.sources?.length ?? 0,
      factsCreated: factRows.length,
      conflictsCreated: conflictRows.length,
      remainingMissingFacts: payload.remainingMissingFacts ?? deal.missingFacts,
    };
  } catch (error) {
    await supabase
      .from("deal_enrichment_runs")
      .update({
        status: "failed",
        error: error instanceof Error ? error.message : "Enrichment failed",
        completed_at: new Date().toISOString(),
      })
      .eq("id", run.id);
    throw error;
  }
}

function parseJson<T>(text: string): T | null {
  try {
    return JSON.parse(text) as T;
  } catch {
    const match = text.match(/\{[\s\S]*\}/);
    if (!match) return null;
    try {
      return JSON.parse(match[0]) as T;
    } catch {
      return null;
    }
  }
}

File: apps/backend-api/app/services/deal_enrichment.py
from __future__ import annotations

import json
import logging
from typing import Any
from urllib.parse import urlparse

from fastapi import HTTPException

from app.repositories.supabase_repo import repo
from app.services.deepseek_llm import llm
from app.services.keiro_search import keiro

logger = logging.getLogger("towerbrook.deal_enrichment")

SYSTEM = """You enrich private-equity deal facts using web search.
Use authoritative sources first: buyer, seller, target, investor, bank, law firm, regulator, reputable trade press.
Never invent undisclosed economics. Return strict JSON only. Low-confidence or uncertain facts must be review-gated."""


async def run_deal_enrichment(external_deal_id: str) -> dict[str, Any]:
    if not repo.enabled or not repo.client:
        raise HTTPException(
            status_code=503,
            detail="Supabase is not configured on the backend API.",
        )
    if not llm.configured:
        raise HTTPException(
            status_code=503,
            detail="DEEPSEEK_API_KEY is not configured on the backend API.",
        )

    client = repo.client
    deal_rows = (
        client.table("deals")
        .select("*")
        .eq("external_id", external_deal_id)
        .limit(1)
        .execute()
        .data
        or []
    )
    deal_row = deal_rows[0] if deal_rows else None
    if not deal_row:
        raise HTTPException(status_code=404, detail=f"Unknown persisted deal: {external_deal_id}")

    facts_result = (
        client.table("deal_facts").select("fact_type,fact_value").eq("deal_id", deal_row["id"]).execute()
    )
    known_facts = facts_result.data or []
    missing_facts = deal_row.get("missing_facts") or []
    follow_up_searches = deal_row.get("follow_up_searches") or []

    run_result = (
        client.table("deal_enrichment_runs")
        .insert(
            {
                "deal_id": deal_row["id"],
                "trigger": "manual",
                "status": "running",
                "queries": follow_up_searches,
                "metadata": {"dealExternalId": external_deal_id},
            }
        )
        .execute()
    )
    run = (run_result.data or [None])[0]
    if not run:
        raise HTTPException(status_code=500, detail="Failed to create enrichment run.")
    run_id = run["id"]

    try:
        search_sources = await _collect_search_evidence(deal_row["name"], follow_up_searches)
        payload = await _extract_enrichment_payload(
            deal_row=deal_row,
            known_facts=known_facts,
            missing_facts=missing_facts,
            follow_up_searches=follow_up_searches,
            search_sources=search_sources,
        )

        source_id_by_url: dict[str, str] = {}
        sources_to_persist = payload.get("sources") or search_sources
        for source in sources_to_persist:
            source_id = _upsert_web_source(client, source, external_deal_id)
            source_id_by_url[source["url"]] = source_id
            snippet = source.get("snippet")
            if snippet:
                client.table("source_chunks").insert(
                    {
                        "source_id": source_id,
                        "content": snippet,
                        "token_count": max(1, len(snippet.split())),
                        "metadata": {
                            "dealExternalId": external_deal_id,
                            "enrichmentRunId": run_id,
                        },
                    }
                ).execute()

        fact_rows = []
        for fact in payload.get("facts") or []:
            source_url = fact.get("sourceUrl")
            source_id = source_id_by_url.get(source_url) if source_url else None
            if not source_id and source_url:
                source_id = _upsert_web_source(
                    client,
                    {
                        "title": source_url,
                        "url": source_url,
                        "snippet": fact.get("evidenceText"),
                    },
                    external_deal_id,
                )
            fact_rows.append(
                {
                    "deal_id": deal_row["id"],
                    "fact_type": fact.get("factType"),
                    "fact_value": fact.get("factValue"),
                    "normalized_value": fact.get("normalizedValue"),
                    "source_id": source_id,
                    "evidence_text": fact.get("evidenceText"),
                    "confidence": max(0.0, min(1.0, float(fact.get("confidence") or 0.7))),
                    "extraction_method": "web_search",
                    "review_status": fact.get("reviewStatus") or "needs_review",
                }
            )

        if fact_rows:
            client.table("deal_facts").insert(fact_rows).execute()

        conflict_rows = [
            {
                "deal_id": deal_row["id"],
                "fact_type": conflict.get("factType"),
                "values": conflict.get("values") or [],
                "note": conflict.get("note") or "",
            }
            for conflict in payload.get("conflicts") or []
        ]
        if conflict_rows:
            client.table("deal_fact_conflicts").insert(conflict_rows).execute()

        remaining_missing = payload.get("remainingMissingFacts") or missing_facts
        client.table("deals").update(
            {"missing_facts": remaining_missing, "updated_at": "now()"}
        ).eq("external_id", external_deal_id).execute()

        client.table("deal_enrichment_runs").update(
            {
                "status": "completed",
                "sources_found": len(sources_to_persist),
                "facts_created": len(fact_rows),
                "completed_at": "now()",
            }
        ).eq("id", run_id).execute()

        return {
            "runId": run_id,
            "sourcesFound": len(sources_to_persist),
            "factsCreated": len(fact_rows),
            "conflictsCreated": len(conflict_rows),
            "remainingMissingFacts": remaining_missing,
        }
    except Exception as exc:
        logger.exception("Deal enrichment failed for %s", external_deal_id)
        client.table("deal_enrichment_runs").update(
            {
                "status": "failed",
                "error": str(exc),
                "completed_at": "now()",
            }
        ).eq("id", run_id).execute()
        if isinstance(exc, HTTPException):
            raise
        raise HTTPException(status_code=500, detail=str(exc)) from exc


async def _collect_search_evidence(deal_name: str, queries: list[str]) -> list[dict[str, str]]:
    seen: dict[str, dict[str, str]] = {}
    selected = (queries or [f"{deal_name} transaction advisor completion"])[:4]
    for query in selected:
        for result in await keiro.search(query, limit=6):
            url = result.get("url")
            if not url or url in seen:
                continue
            seen[url] = {
                "title": result.get("title") or url,
                "url": url,
                "publisher": _safe_publisher(url),
                "snippet": result.get("content") or result.get("snippet") or "",
            }
    return list(seen.values())[:12]


async def _extract_enrichment_payload(
    *,
    deal_row: dict[str, Any],
    known_facts: list[dict[str, Any]],
    missing_facts: list[str],
    follow_up_searches: list[str],
    search_sources: list[dict[str, str]],
) -> dict[str, Any]:
    known_fact_lines = "\n".join(
        f"- {fact['fact_type']}: {fact['fact_value']}" for fact in known_facts
    )
    source_blocks = "\n\n".join(
        f"{index + 1}. {source['title']}\nURL: {source['url']}\nPublisher: {source.get('publisher', '')}\nSnippet: {source.get('snippet', '')}"
        for index, source in enumerate(search_sources)
    )
    user = f"""Deal: {deal_row['name']}
Theme: {deal_row.get('theme')}
Known facts:
{known_fact_lines or '- none'}

Missing facts:
{chr(10).join(f'- {fact}' for fact in missing_facts) or '- none'}

Run targeted searches using these query ideas:
{chr(10).join(f'- {query}' for query in follow_up_searches) or '- none'}

Source snippets from backend web search:
{source_blocks or 'No live search results returned.'}

Return strict JSON:
{{
  "sources": [{{"title": string, "url": string, "publisher": string, "snippet": string}}],
  "facts": [{{"factType": string, "factValue": string, "normalizedValue": string, "sourceUrl": string, "evidenceText": string, "confidence": number, "reviewStatus": "verified" | "needs_review" | "not_disclosed"}}],
  "conflicts": [{{"factType": string, "values": string[], "note": string}}],
  "remainingMissingFacts": string[]
}}"""
    raw = await llm.complete(SYSTEM, user, max_tokens=3000, json_mode=True)
    return _parse_json(raw) or {}


def _upsert_web_source(client: Any, source: dict[str, str], external_deal_id: str) -> str:
    record = repo.upsert_source(
        {
            "title": source.get("title") or source.get("url") or "Web source",
            "url": source.get("url"),
            "publisher": source.get("publisher"),
            "source_type": "web_enrichment",
            "raw_text": source.get("snippet"),
            "metadata": {"dealExternalId": external_deal_id},
        }
    )
    return record.id


def _parse_json(text: str) -> dict[str, Any] | None:
    try:
        parsed = json.loads(text)
        return parsed if isinstance(parsed, dict) else None
    except json.JSONDecodeError:
        start = text.find("{")
        end = text.rfind("}")
        if start < 0 or end <= start:
            return None
        try:
            parsed = json.loads(text[start : end + 1])
            return parsed if isinstance(parsed, dict) else None
        except json.JSONDecodeError:
            return None


def _safe_publisher(url: str) -> str:
    try:
        hostname = urlparse(url).hostname
        return hostname.replace("www.", "") if hostname else "Web source"
    except Exception:
        return "Web source"
